File: packages/host-core/src/view-recorder.ts
import type { ComposeTrace } from "@kohaku-ui/composer";
import type { JsonObject, Surface, UISpec } from "@kohaku-ui/spec-core";
import { failOpen } from "./errors.js";

/**
 * Recording hooks for View Lineage (@kohaku-ui/lineage supplies the implementation; no recording if unset).
 * Shared by both host profiles (REST's KohakuHostDeps.recorder and the MCP profile's McpHostDeps.recorder) so
 * a single interface governs what "audit-recording symmetry between the two surfaces" means. Originally
 * defined in host-rest's types.ts; moved here so the MCP profile can depend on the same contract instead of
 * carrying only the narrower `onComposed` callback it used to have.
 */
export interface ViewRecorder {
  composed(args: {
    spec: UISpec;
    trace: ComposeTrace;
    surface: Surface;
    sessionId?: string;
    tenant?: string;
    /** Precomputed hash (shared with the done event so specHash is computed once per request). If passed, the recorder implementation does not re-hash the Spec. If unset, computed internally (backward compatible). */
    specHash?: string;
    structureHash?: string;
  }): Promise<void>;
  interacted(args: {
    intentHash: string;
    specHash?: string;
    componentId: string;
    on: string;
    payload: JsonObject;
    surface: Surface;
    sessionId?: string;
    tenant?: string;
  }): Promise<void>;
  rendered?(args: {
    specHash: string;
    surface: Surface;
    renderer: string;
    durationMs?: number;
    tenant?: string;
  }): Promise<void>;
  componentUsed?(args: {
    artifactId: string;
    surface: Surface;
    outcome: "ok" | "error";
    sessionId?: string;
    tenant?: string;
  }): Promise<void>;
  /**
   * Records a fallback (deterministic downgrade on L1/L2 generation failure / component downgrade via capability
   * negotiation). The judgment source is spec.provenance.fallback. No recording if unset.
   */
  fallback?(args: {
    spec: UISpec;
    reason: string;
    kind: "generation" | "negotiation";
    surface: Surface;
    sessionId?: string;
    tenant?: string;
    /** Precomputed hash. If passed, the recorder implementation does not re-hash the Spec. If unset, computed internally (backward compatible). */
    specHash?: string;
  }): Promise<void>;
}

/**
 * The cancelled-aware, fail-open "record the composed result" sequence shared by both host profiles (REST's
 * deliverComposed/finishStream and the MCP profile's composeAndAudit): a cancelled compose (the caller's
 * abort fired) is not a generation failure and observer.onError already received phase:"cancelled" from the
 * composer, so `record` is skipped entirely — a client disconnect/timeout must not inflate view.composed /
 * view.fallback counts. Otherwise `record` (the host's own composed -> fallback recording, in that order) runs
 * fail-open (host-core's failOpen): a recording failure must not take down an otherwise-successful delivery,
 * and is instead reported to `onError`. Caveat: this only protects `record`; `onError` itself must not throw
 * (a throwing `onError` is not caught here and would escape to the caller).
 */
export async function recordComposedResult(
  result: { spec: UISpec; trace: ComposeTrace },
  record: () => Promise<void>,
  onError: (e: unknown) => void | Promise<void>,
): Promise<void> {
  if (result.trace.cancelled === true) return;
  await failOpen(record, async (e) => {
    await onError(e);
  });
}

/**
 * Records view.fallback when the spec includes a fallback (deterministic downgrade on generation failure /
 * capability-negotiation downgrade). The judgment source is the spec, not the compose trace: negotiation
 * downgrade happens every time at finish even after a cache hit, so without looking at
 * spec.provenance.fallback the downgrade on the cache-hit path would be missed. A missing `kind` is treated as
 * "generation" (compatible with older records that predate the field).
 *
 * Shared by both host profiles (REST's recordFallbackIfAny and the MCP profile's composeAndAudit) so the
 * fallback-detection rule lives in exactly one place instead of two independently-drifting copies.
 */
export async function recordViewFallback(
  recorder: ViewRecorder | undefined,
  spec: UISpec,
  meta: { surface: Surface; sessionId?: string; tenant?: string; specHash?: string },
): Promise<void> {
  const fallback = spec.provenance.fallback;
  if (fallback == null) return;
  await recorder?.fallback?.({
    spec,
    reason: fallback.reason,
    kind: fallback.kind ?? "generation",
    surface: meta.surface,
    ...(meta.specHash != null ? { specHash: meta.specHash } : {}),
    ...(meta.sessionId != null ? { sessionId: meta.sessionId } : {}),
    ...(meta.tenant != null ? { tenant: meta.tenant } : {}),
  });
}

File: python/kohaku/src/kohaku/host_core/view_recorder.py
"""kohaku.host_core.view_recorder — records view.fallback when a composed Spec carries a fallback (port of
packages/host-core/src/view-recorder.ts's `recordViewFallback`).

TS's `view-recorder.ts` additionally defines the shared `ViewRecorder` interface (`composed` / `interacted` /
`rendered?` / `componentUsed?` / `fallback?`). This port only moves `recordViewFallback` itself — both Python
hosts already carry their own structurally-equivalent `ViewRecorderProtocol` (kohaku.host_rest.deps,
kohaku.host_mcp.types), which host_core must not import (host_core may not depend on host_rest/host_mcp per
the layer contract), so `record_view_fallback` below is typed against a narrow, private, fallback-only
Protocol that both hosts' `ViewRecorderProtocol` already structurally satisfy.
"""

from __future__ import annotations

from typing import Protocol

from kohaku.spec import Surface, UISpec


class _FallbackRecorder(Protocol):
    """The narrow slice of ViewRecorder record_view_fallback needs — both host_rest's and host_mcp's own
    `ViewRecorderProtocol.fallback` structurally satisfy this as-is."""

    async def fallback(
        self,
        *,
        spec: UISpec,
        reason: str,
        kind: str,
        surface: Surface,
        session_id: str | None = ...,
        tenant: str | None = ...,
    ) -> None: ...


async def record_view_fallback(
    recorder: _FallbackRecorder | None,
    spec: UISpec,
    *,
    surface: Surface,
    session_id: str | None = None,
    tenant: str | None = None,
) -> None:
    """Records view.fallback when the spec includes a fallback (deterministic downgrade on L1/L2 generation
    failure / component downgrade via capability negotiation). The judgment source is `spec.provenance.fallback`,
    not the compose trace: negotiation downgrade happens every time at finish even after a cache hit, so
    without looking at `spec.provenance.fallback` the downgrade on the cache-hit path would be missed. A
    missing `kind` is treated as "generation" (compatible with older records that predate the field).

    Shared by both host profiles (REST's record_fallback_if_any and the MCP profile's `_audit_compose`) so the
    fallback-detection rule lives in exactly one place instead of two independently-drifting copies. No-op
    when `recorder` is None (matching both hosts' existing "recording is optional" contract).
    """
    fallback = spec.provenance.fallback
    if fallback is None:
        return
    if recorder is None:
        return
    await recorder.fallback(
        spec=spec,
        reason=fallback.reason,
        kind=fallback.kind if fallback.kind is not None else "generation",
        surface=surface,
        session_id=session_id,
        tenant=tenant,
    )
